rfcs/20180920-unify-rnn-interface.md

```diff
@@ -0,0 +1,377 @@
+# Unify RNN Inteface
+
+| Status        | Proposed                                             |
+| ------------- |:---------------------------------------------------- |
+| **Author(s)** | @qlzh727 (scottzhu@google.com)                       |
+| **Sponsor**   | @ebrevdo (ebrevdo@google.com), @drpng (drpng@google.com)|
+| **Updated**   | 2018-08-24                                           |
+
+## Objective
+
+Unify RNN (LSTM/GRU/other recurrent models) interfaces between TF RNN and Keras RNN at the release 
+of TensorFlow 2.0.
+
+## Motivation
+
+Recurrent neural networks [RNN](https://en.wikipedia.org/wiki/Recurrent_neural_network) is the deep 
+neural network that works great with sequential input data like text and voice. There are two sets 
+of APIs in TensorFlow that allows users to build a RNN. The APIs are a bit different from each 
+other, but trying to solve the same problem. This situation creates some user confusions about:
+
+1. What's the recommended API to use.
+1. What's the relationship between those 2 APIs and how they work together (which they don't 
+currently).
+
+Ideally those two APIs should be unified into "one true API", which should be:
+1. Intuitive to use (hide unnecessary complexity from user).
+1. Covers most of the common use cases.
+1. Easy to migrate from either side.
+
+Keras has become the recommended high level API for Tensorflow since its easy to understand and use.
+We would prefer to unify the final API that is similar to existing Keras API, and port 
+functionalities from TF RNN to Keras.
+
+
+## Design Proposal
+
+1. Update RNN cells in Keras and TF RNN, so that all of them will work with Keras RNN layer and TF 
+RNN layer.
+1. Remove the duplicated RNN cells between Keras and TF, update the documentation for the migration 
+path.
+    1. tf.nn.rnn_cell.BasicRNNCell -> tf.keras.SimpleRNNCell
+    1. tf.nn.rnn_cell.BasicLSTMCell,  tf.keras.LSTMCell -> tf.nn.rnn_cell.LSTMCell
+    1. tf.nn.rnn_cell.GRUCell -> tf.keras.GRUCell
+    1. tf.nn.rnn_cell.MultiRNNCell -> tf.keras.StackedRNNCells
+1. Move existing tf.nn.rnn* code from tensorflow/python/ops/ to tensorflow/python/rnn. The RNN layer
+and cell are not tensorflow ops, and is a bit weird to stay under ops. Externally the tf.export will
+ensure the object/function stays the same.
+1. Migrate selected RNN cell from contrib to core (tf/python/rnn). See detailed section below.
+1. Still keep both TF and Keras RNN layer API in TF 2.0, since both APIs are heavily used. Update 
+documentation to state that Keras is the preferred API to use, and encourage user to use Keras. Add 
+warning message in TF RNN layer, point user to documentation for suggested high level RNN layer API.
```

Comment:
Could you clarify how the gates ordering will be impacted? Currently Keras and TF weights have different orders for both LSTM and GRU (e.g. for LSTM, TF has ICFO ordering when Keras has IFCO). Will those be unified or will they remain unchanged?

Reply:
I think those will remain the same since they are just implementation details. The weights between Keras and TF cells has other differences than that, so we don't expect the weights to work between them.

Reply:
I agree, these are implementation details. However, if you keep the gates ordering different in the two, this can imply a performance difference between the two (in particular when an acceleration library is used underneath, e.g. MKL-DNN or CUDNN).
From a usability point of view, it can be odd to have 2 interface layers with similar functionality but different performance.

Reply:
Sorry for the late reply. 

I see what you are trying to address here. I think there is a issue like circular dependency here: the underlying acceleration library could like a stable input format from TF, and TF also would like to have a unified format from all the underlying library. 

Also I am not sure whether there will be a performance implication for this reason. The weights of the individual gate should be saved separately, we could pass them to underlying library with different order, it does not necessary mean that we have to break them and restitch back. Can u clarify the reason for the performance part a bit more?

Reply:
Thanks for the reply.

The performance concern is exactly because of the weights reorders as you mention.
In the cases where the number of iterations is small (think seq2seq decoder), you either will need shadow copies of the weights that are reused from iteration to iteration, or you will have to reorder for each call to the library, which can have a large overhead (especially for inference).

Saving them separately would not solve the issue as all libraries I know of require to have the gate weights packed into one big matrix (it enables to run one big GEMM instead of many smaller ones). So you would still have some copy overhead.

From a library point of view, we are fine with whatever format you pick. However, if you have 2 formats in the same framework, it becomes harder to accommodate efficiently.

Reply:
Ack, we will have discussion internally about the gate format. I think most of the LSTM cell in TF are in ICFO order, except the cudnn LSTM cell, which is IFCO since the underlying cudnn lib require the input to be in that format. I think unifying into one format will be helpful. Will post the discussion outcome here within this week.

Reply:
Thanks. As mentioned earlier, it seems that [tf.keras uses IFCO](https://github.com/keras-team/keras/blob/f534cb0c459ce2a8384e3bc0375eb1cbf80edb56/keras/layers/recurrent.py#L1899) too.